File: templates/problems/06_linked_list/06_copy_random_list/python/solution.py
from helpers import RandomNode, Optional


def solve_copy_random_list(head: Optional[RandomNode]) -> Optional[RandomNode]:
    """Create a deep copy using a hash map.

    Time: O(n), Space: O(n).
    """
    if head is None:
        return None
    old_to_new: dict[int, RandomNode] = {}
    curr = head
    while curr is not None:
        old_to_new[id(curr)] = RandomNode(val=curr.val)
        curr = curr.next
    curr = head
    while curr is not None:
        node = old_to_new[id(curr)]
        if curr.next is not None:
            node.next = old_to_new[id(curr.next)]
        if curr.random is not None:
            node.random = old_to_new[id(curr.random)]
        curr = curr.next
    return old_to_new[id(head)]

File: templates/problems/06_linked_list/06_copy_random_list/typescript/solution.ts
import { RandomNode } from "./helpers";

export function solveCopyRandomList(head: RandomNode | null): RandomNode | null {
    if (head === null) {
        return null;
    }
    const oldToNew = new Map<RandomNode, RandomNode>();
    let curr: RandomNode | null = head;
    while (curr !== null) {
        oldToNew.set(curr, new RandomNode(curr.val));
        curr = curr.next;
    }
    curr = head;
    while (curr !== null) {
        const node = oldToNew.get(curr)!;
        if (curr.next !== null) {
            node.next = oldToNew.get(curr.next)!;
        }
        if (curr.random !== null) {
            node.random = oldToNew.get(curr.random)!;
        }
        curr = curr.next;
    }
    return oldToNew.get(head)!;
}
